File: Server/src/services/mpesaService.js
const axios = require('axios');
const crypto = require('crypto');
const env = require('../config/env');
const logger = require('../utils/logger');
const { Pool } = require('pg');

// Database pool
const pool = new Pool({
  connectionString: process.env.DATABASE_URL || 'postgresql://postgres:password@localhost:5432/emergency_wallet'
});

// Demo mode flag - set to true for development/testing
const DEMO_MODE = process.env.MPESA_DEMO_MODE !== 'false';

/**
 * Generate M-Pesa transaction reference
 */
const generateMpesaReference = () => {
  const timestamp = Date.now();
  const random = Math.floor(Math.random() * 10000).toString().padStart(4, '0');
  return `MJ${timestamp.toString().slice(-6)}${random}`;
};

/**
 * Simulate C2B (Customer to Business) transaction
 */
const simulateC2B = ({ paybillNumber, accountNumber, amount }) => {
  if (!DEMO_MODE) return;

  const reference = generateMpesaReference();
  logger.info(`[M-Pesa Demo] C2B Simulation:`, {
    paybillNumber,
    accountNumber,
    amount,
    reference
  });

  // Simulate SMS confirmation
  setTimeout(() => {
    logger.info(`[M-Pesa Demo] SMS: KES ${amount} paid to ${paybillNumber} Account ${accountNumber}. Ref: ${reference}`);
  }, 2000);
};

/**
 * Simulate B2C (Business to Customer) transaction
 */
const simulateB2C = ({ tillNumber, amount }) => {
  if (!DEMO_MODE) return;

  const reference = generateMpesaReference();
  logger.info(`[M-Pesa Demo] B2C Simulation:`, {
    tillNumber,
    amount,
    reference
  });

  // Simulate SMS confirmation
  setTimeout(() => {
    logger.info(`[M-Pesa Demo] SMS: KES ${amount} paid to Till ${tillNumber}. Ref: ${reference}`);
  }, 2000);
};

/**
 * Simulate STK Push transaction
 */
const simulateSTKPush = ({ phoneNumber, amount }) => {
  if (!DEMO_MODE) return;

  const reference = generateMpesaReference();
  logger.info(`[M-Pesa Demo] STK Push Simulation:`, {
    phoneNumber,
    amount,
    reference
  });

  // Simulate STK push prompt
  setTimeout(() => {
    logger.info(`[M-Pesa Demo] STK Push: Enter PIN to send KES ${amount} to ${phoneNumber}`);
  }, 1000);

  // Simulate SMS confirmation
  setTimeout(() => {
    logger.info(`[M-Pesa Demo] SMS: KES ${amount} sent to ${phoneNumber}. Ref: ${reference}`);
  }, 3000);
};

class MpesaService {
  constructor() {
    this.consumerKey = env.MPESA_CONSUMER_KEY;
    this.consumerSecret = env.MPESA_CONSUMER_SECRET;
    this.passkey = env.MPESA_PASSKEY;
    this.shortcode = env.MPESA_SHORTCODE;
    this.environment = env.MPESA_ENV;
    this.baseURL = this.environment === 'production'
      ? 'https://api.safaricom.co.ke'
      : 'https://sandbox.safaricom.co.ke';
    this.accessToken = null;
    this.tokenExpiry = null;
  }

  async getAccessToken() {
    if (this.accessToken && this.tokenExpiry > Date.now()) return this.accessToken;
    const auth = Buffer.from(`${this.consumerKey}:${this.consumerSecret}`).toString('base64');
    const response = await axios.get(`${this.baseURL}/oauth/v1/generate?grant_type=client_credentials`, {
      headers: { Authorization: `Basic ${auth}` },
    });
    this.accessToken = response.data.access_token;
    this.tokenExpiry = Date.now() + (response.data.expires_in - 60) * 1000;
    return this.accessToken;
  }

  async stkPush(phoneNumber, amount) {
    // Check if we're in test mode (no real credentials)
    if (!this.consumerKey || !this.consumerSecret || this.consumerKey === 'your_consumer_key') {
      logger.info('Using mock STK Push for testing');
      return {
        CheckoutRequestID: `MOCK_${Date.now()}`,
        MerchantRequestID: `MOCK_MERCHANT_${Date.now()}`,
        ResponseCode: '0',
        ResponseDescription: 'Success. Request accepted for processing',
        CustomerMessage: 'Mock STK Push initiated successfully'
      };
    }

    const token = await this.getAccessToken();
    const timestamp = new Date().toISOString().replace(/[^0-9]/g, '').slice(0, 14);
    const password = Buffer.from(`${this.shortcode}${this.passkey}${timestamp}`).toString('base64');

    const data = {
      BusinessShortCode: this.shortcode,
      Password: password,
      Timestamp: timestamp,
      TransactionType: 'CustomerPayBillOnline',
      Amount: amount,
      PartyA: phoneNumber,
      PartyB: this.shortcode,
      PhoneNumber: phoneNumber,
      CallBackURL: `${env.CALLBACK_BASE_URL || 'https://your-domain.com'}/api/mpesa/callback`,
      AccountReference: 'EmergencyWallet',
      TransactionDesc: 'Wallet recharge',
    };

    const response = await axios.post(`${this.baseURL}/mpesa/stkpush/v1/processrequest`, data, {
      headers: { Authorization: `Bearer ${token}` },
    });
    logger.info('STK Push response', response.data);
    return response.data;
  }

  async b2c(phoneNumber, amount, transactionId) {
    const token = await this.getAccessToken();
    const data = {
      InitiatorName: 'testapi',
      SecurityCredential: 'placeholder_encrypted_credential', // would require actual encryption
      CommandID: 'BusinessPayment',
      Amount: amount,
      PartyA: this.shortcode,
      PartyB: phoneNumber,
      Remarks: `EmergencyWallet payment ${transactionId}`,
      QueueTimeOutURL: `${env.CALLBACK_BASE_URL}/mpesa/timeout`,
      ResultURL: `${env.CALLBACK_BASE_URL}/mpesa/b2c-result`,
      Occasion: 'Payment',
    };
    // In sandbox you may mock this call
    logger.warn('B2C not fully implemented – mock response');
    return { ConversationID: 'mock', ResponseCode: '0', ResponseDescription: 'Success' };
  }
}

module.exports = new MpesaService();

// Export additional functions for frontend compatibility
module.exports.DEMO_MODE = DEMO_MODE;
module.exports.simulateC2B = simulateC2B;
module.exports.simulateB2C = simulateB2C;
module.exports.simulateSTKPush = simulateSTKPush;
module.exports.generateMpesaReference = generateMpesaReference;

File: Backend/EmergencyWallet/src/main/java/org/NovaGroup/EmergencyWallet/Mpesa/MpesaService.java
package org.NovaGroup.EmergencyWallet.Mpesa;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import okhttp3.*;
import org.NovaGroup.EmergencyWallet.Mpesa.models.MpesaTokenResponse;
import org.NovaGroup.EmergencyWallet.Mpesa.models.StkPushResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

@Service
@Slf4j
public class MpesaService {

    @Value("${mpesa.consumer-key}")
    private String consumerKey;

    @Value("${mpesa.consumer-secret}")
    private String consumerSecret;

    @Value("${mpesa.shortcode}")
    private String shortcode;

    @Value("${mpesa.passkey}")
    private String passkey;

    @Value("${mpesa.callback-url}")
    private String callbackUrl;

    @Value("${mpesa.stk-callback-url}")
    private String stkCallbackUrl;

    @Value("${mpesa.env:sandbox}")
    private String environment;

    @Value("${demo.mode:true}")
    private boolean demoMode;

    private static final String SANDBOX_AUTH_URL = "https://sandbox.safaricom.co.ke/oauth/v1/generate?grant_type=client_credentials";
    private static final String PROD_AUTH_URL = "https://api.safaricom.co.ke/oauth/v1/generate?grant_type=client_credentials";
    private static final String SANDBOX_STK_URL = "https://sandbox.safaricom.co.ke/mpesa/stkpush/v1/processrequest";
    private static final String PROD_STK_URL = "https://api.safaricom.co.ke/mpesa/stkpush/v1/processrequest";
    private static final String SANDBOX_C2B_URL = "https://sandbox.safaricom.co.ke/mpesa/c2b/v1/simulate";
    private static final String PROD_C2B_URL = "https://api.safaricom.co.ke/mpesa/c2b/v1/simulate";
    private static final String SANDBOX_B2C_URL = "https://sandbox.safaricom.co.ke/mpesa/b2c/v1/paymentrequest";
    private static final String PROD_B2C_URL = "https://api.safaricom.co.ke/mpesa/b2c/v1/paymentrequest";
    private static final String SANDBOX_B2B_URL = "https://sandbox.safaricom.co.ke/mpesa/b2b/v1/paymentrequest";
    private static final String PROD_B2B_URL = "https://api.safaricom.co.ke/mpesa/b2b/v1/paymentrequest";

    private String accessToken;
    private LocalDateTime tokenExpiry;
    private final OkHttpClient client = new OkHttpClient();
    private final ObjectMapper objectMapper = new ObjectMapper();

    public String getAccessToken() throws IOException {
        if (accessToken != null && tokenExpiry != null && LocalDateTime.now().isBefore(tokenExpiry)) {
            log.debug("Returning cached access token");
            return accessToken;
        }

        String authUrl = "sandbox".equalsIgnoreCase(environment) ? SANDBOX_AUTH_URL : PROD_AUTH_URL;
        String credentials = Base64.getEncoder().encodeToString((consumerKey + ":" + consumerSecret).getBytes(StandardCharsets.UTF_8));

        log.debug("Encoded credentials: {}", credentials);

        Request request = new Request.Builder()
                .url(authUrl)
                .header("Authorization", "Basic " + credentials)
                .header("User-Agent", "Java/OkHttp")
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Failed to authenticate with Daraja API. HTTP: " + response.code());
            }

            String responseBodyStr = response.body() != null ? response.body().string() : "";
            log.debug("Token Response: {}", responseBodyStr);

            MpesaTokenResponse tokenResponse = objectMapper.readValue(responseBodyStr, MpesaTokenResponse.class);
            accessToken = tokenResponse.getAccess_token();
            tokenExpiry = LocalDateTime.now().plusMinutes(55);

            log.info("M-Pesa access token obtained successfully");
            log.debug("Access Token: {}", accessToken);
            return accessToken;
        }
    }

    public StkPushResponse initiateStkPush(String phoneNumber, long amount, String accountReference) throws IOException {
        if (demoMode) {
            Map<String, Object> simulated = simulateStkPush(phoneNumber, amount, accountReference);
            // map simulated response to DTO if needed
            return objectMapper.convertValue(simulated, StkPushResponse.class);
        }

        String token = getAccessToken();
        String timestamp = generateTimestamp();
        String password = generatePassword(timestamp);
        String url = "sandbox".equalsIgnoreCase(environment) ? SANDBOX_STK_URL : PROD_STK_URL;

        Map<String, Object> payload = new HashMap<>();
        payload.put("BusinessShortCode", shortcode);
        payload.put("Password", password);
        payload.put("Timestamp", timestamp);
        payload.put("TransactionType", "CustomerPayBillOnline");
        payload.put("Amount", amount);
        payload.put("PartyA", phoneNumber);
        payload.put("PartyB", shortcode);
        payload.put("PhoneNumber", phoneNumber);
        payload.put("CallBackURL", stkCallbackUrl);
        payload.put("AccountReference", accountReference);
        payload.put("TransactionDesc", "Emergency Wallet Deposit");

        RequestBody body = RequestBody.create(
                objectMapper.writeValueAsString(payload),
                MediaType.parse("application/json")
        );

        Request request = new Request.Builder()
                .url(url)
                .header("Authorization", "Bearer " + token)
                .header("User-Agent", "Java/OkHttp")
                .post(body)
                .build();

        try (Response response = client.newCall(request).execute()) {
            String responseBodyStr = response.body() != null ? response.body().string() : "";
            log.debug("STK Push response body: {}", responseBodyStr);

            StkPushResponse stkResponse = objectMapper.readValue(responseBodyStr, StkPushResponse.class);
            log.info("STK Push initiated. Response code: {}", stkResponse.getResponseCode());
            return stkResponse;
        }
    }
    public Map<String, Object> simulateStkPush(String phoneNumber, long amount, String accountReference) {
        log.info("[DEMO MODE] Simulating STK Push. Phone: {}, Amount: {}", phoneNumber, amount);

        Map<String, Object> response = new HashMap<>();
        response.put("MerchantRequestID", "ws_CO_" + System.currentTimeMillis());
        response.put("CheckoutRequestID", "ws_CO_" + System.currentTimeMillis());
        response.put("ResponseCode", "0");
        response.put("ResponseDescription", "Success. Request accepted for processing.");
        response.put("CustomerMessage", "Success. Request accepted for processing.");

        return response;
    }

    public Map<String, Object> processC2B(String paybillNumber, String accountNumber, String phoneNumber, long amount) throws IOException {
        if (demoMode) {
            return simulateC2B(paybillNumber, accountNumber, amount);
        }

        String token = getAccessToken();
        String url = "sandbox".equalsIgnoreCase(environment) ? SANDBOX_C2B_URL : PROD_C2B_URL;

        Map<String, Object> payload = new HashMap<>();
        payload.put("ShortCode", paybillNumber);
        payload.put("CommandID", "CustomerPayBillOnline");
        payload.put("Amount", amount);
        payload.put("Msisdn", phoneNumber);
        payload.put("BillRefNumber", accountNumber);

        RequestBody body = RequestBody.create(
            objectMapper.writeValueAsString(payload),
            MediaType.parse("application/json")
        );

        Request request = new Request.Builder()
            .url(url)
            .header("Authorization", "Bearer " + token)
            .post(body)
            .build();

        try (Response response = client.newCall(request).execute()) {
            Map<String, Object> responseBody = objectMapper.readValue(response.body().string(), Map.class);
            log.info("C2B Paybill payment initiated. Response: {}", responseBody.get("ResponseCode"));
            return responseBody;
        }
    }

    public Map<String, Object> simulateC2B(String paybillNumber, String accountNumber, long amount) {
        log.info("[DEMO MODE] Simulating C2B. Paybill: {}, Account: {}, Amount: {}", 
            paybillNumber, accountNumber, amount);
        
        Map<String, Object> response = new HashMap<>();
        response.put("ConversationID", "c2b_" + System.currentTimeMillis());
        response.put("OriginatorConversationID", "oc2b_" + System.currentTimeMillis());
        response.put("ResponseCode", "0");
        response.put("ResponseDescription", "Accept the service request successfully.");
        
        return response;
    }

    public Map<String, Object> processB2B(String receiverShortcode, long amount, String accountReference) throws IOException {
        if (demoMode) {
            return simulateB2B(receiverShortcode, amount);
        }

        String token = getAccessToken();
        String timestamp = generateTimestamp();
        String url = "sandbox".equalsIgnoreCase(environment) ? SANDBOX_B2B_URL : PROD_B2B_URL;

        Map<String, Object> payload = new HashMap<>();
        payload.put("Initiator", shortcode);
        payload.put("SecurityCredential", generateSecurityCredential(shortcode));
        payload.put("CommandID", "BusinessPayBill");
        payload.put("SenderIdentifierType", "4"); // Business till/Account
        payload.put("RecieverIdentifierType", "4");
        payload.put("Amount", amount);
        payload.put("PartyA", shortcode);
        payload.put("PartyB", receiverShortcode);
        payload.put("Remarks", "B2B Payment");
        payload.put("QueueTimeOutURL", callbackUrl);
        payload.put("ResultURL", callbackUrl);
        payload.put("AccountReference", accountReference);

        RequestBody body = RequestBody.create(
            objectMapper.writeValueAsString(payload),
            MediaType.parse("application/json")
        );

        Request request = new Request.Builder()
            .url(url)
            .header("Authorization", "Bearer " + token)
            .post(body)
            .build();

        try (Response response = client.newCall(request).execute()) {
            Map<String, Object> responseBody = objectMapper.readValue(response.body().string(), Map.class);
            log.info("B2B payment initiated. Response: {}", responseBody.get("ResponseCode"));
            return responseBody;
        }
    }

    public Map<String, Object> processB2C(String phoneNumber, long amount, String commandId, String remarks) throws IOException {
        if (demoMode) {
            return simulateB2C(phoneNumber, amount);
        }

        String token = getAccessToken();
        String url = "sandbox".equalsIgnoreCase(environment) ? SANDBOX_B2C_URL : PROD_B2C_URL;

        Map<String, Object> payload = new HashMap<>();
        payload.put("Initiator", shortcode);
        payload.put("SecurityCredential", generateSecurityCredential(shortcode));
        payload.put("CommandID", commandId); // SalaryPayment, BusinessPayment, PromotionPayment
        payload.put("Amount", amount);
        payload.put("PartyA", shortcode);
        payload.put("PartyB", phoneNumber);
        payload.put("Remarks", remarks != null ? remarks : "B2C Payment");
        payload.put("QueueTimeOutURL", callbackUrl);
        payload.put("ResultURL", callbackUrl);

        RequestBody body = RequestBody.create(
            objectMapper.writeValueAsString(payload),
            MediaType.parse("application/json")
        );

        Request request = new Request.Builder()
            .url(url)
            .header("Authorization", "Bearer " + token)
            .post(body)
            .build();

        try (Response response = client.newCall(request).execute()) {
            Map<String, Object> responseBody = objectMapper.readValue(response.body().string(), Map.class);
            log.info("B2C payment initiated. Response: {}", responseBody.get("ResponseCode"));
            return responseBody;
        }
    }

    public Map<String, Object> simulateB2B(String receiverShortcode, long amount) {
        log.info("[DEMO MODE] Simulating B2B. Receiver: {}, Amount: {}", receiverShortcode, amount);

        Map<String, Object> response = new HashMap<>();
        response.put("ConversationID", "b2b_" + System.currentTimeMillis());
        response.put("OriginatorConversationID", "ob2b_" + System.currentTimeMillis());
        response.put("ResponseCode", "0");
        response.put("ResponseDescription", "Accept the service request successfully.");

        return response;
    }

    public Map<String, Object> simulateB2C(String phoneNumber, long amount) {
        log.info("[DEMO MODE] Simulating B2C. Phone: {}, Amount: {}", phoneNumber, amount);

        Map<String, Object> response = new HashMap<>();
        response.put("ConversationID", "b2c_" + System.currentTimeMillis());
        response.put("OriginatorConversationID", "ob2c_" + System.currentTimeMillis());
        response.put("ResponseCode", "0");
        response.put("ResponseDescription", "Accept the service request successfully.");

        return response;
    }

    private String generateTimestamp() {
        return LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMddHHmmss"));
    }

    private String generatePassword(String timestamp) {
        String rawPassword = shortcode + passkey + timestamp;
        return Base64.getEncoder().encodeToString(rawPassword.getBytes(StandardCharsets.UTF_8));
    }

    private String generateSecurityCredential(String shortcode) {
        // In production, this should use actual encryption
        // For now, return base64 encoded shortcode as placeholder
        return Base64.getEncoder().encodeToString(shortcode.getBytes(StandardCharsets.UTF_8));
    }
}
